biggerquery/interactive.py

```diff
@@ -0,0 +1,785 @@
+        """Returns the result of the specified operation in the form of the pandas.DataFrame, without really running the
+        operation and affecting the table."""
+        not_none_or_error(runtime, 'runtime')
+        not_none_or_error(operation_name, 'operation_name')
+        not_none_or_error(limit, 'limit')
+        results_container, component_callable = decorate_component_dependencies_with_operation_level_dataset_manager(
+            self._standard_component, operation_name=operation_name, peek=True, peek_limit=limit)
+        Job(component_callable, **self._dependency_config).run(runtime)
+        try:
+            return results_container[0]
+        except IndexError:
+            if operation_name != DEFAULT_OPERATION_NAME:
+                raise ValueError("Operation '{}' not found".format(operation_name))
+            else:
+                raise ValueError("You haven't specified operation_name.".format(operation_name))
+
+    def __call__(self, **kwargs):
+        _, component_callable = decorate_component_dependencies_with_operation_level_dataset_manager(self._standard_component)
+        return component_callable(**kwargs)
+
+
+def decorate_component_dependencies_with_operation_level_dataset_manager(
+        standard_component,
+        operation_name=None,
+        peek=None,
+        peek_limit=None):
+    operation_settings = {'operation_name': operation_name, 'peek': peek, 'peek_limit': peek_limit}
+    results_container = []
+
+    def component_callable(**kwargs):
+        operation_level_dataset_managers = {k: OperationLevelDatasetManager(v, **operation_settings)
+                                            for k, v in kwargs.items()}
+
+        component_return_value = standard_component(**operation_level_dataset_managers)
+
+        for operation_level_dataset_manager in operation_level_dataset_managers.values():
+            if operation_level_dataset_manager._result:
+                results_container.extend(operation_level_dataset_manager._result)
+
+        return component_return_value
+
+    component_callable_with_proper_signature = "def reworked_function({signature}):\n    return original_func({kwargs})\n".format(
+        signature=','.join([arg for arg in getargspec(standard_component).args]),
+        kwargs=','.join(['{arg}={arg}'.format(arg=arg) for arg in getargspec(standard_component).args]))
+    component_callable_with_proper_signature_code = compile(component_callable_with_proper_signature, "fakesource",
+                                                            "exec")
+    fake_globals = {}
+    eval(component_callable_with_proper_signature_code, {'original_func': component_callable}, fake_globals)
+
+    component_callable = fake_globals['reworked_function']
+    component_callable.__name__ = operation_name or standard_component.__name__
+
+    return results_container, component_callable
+
+
+class OperationLevelDatasetManager(object):
```

Comment:
I think this class redundant. We could move this logic to `InteractiveDatasetManager`

Reply:
Those classes exist for completely different reasons. OperationLevelDatasetManager has nothing to do with InteractiveDatasetManager, it could be merged with InteractiveComponent, but it would produce super complex class.